### DISCUSSION TITLE
Asynchronous List call / Csv generation and Mail Service

### DISCUSSION THREAD

**Discussion by kilroyFR** [serenity-is/Serenity#6355] [Help]
Hey all, i noticed that when using large table and user queries to EXPORT to EXCEL functionality it can take a lot of time and cause a freeze leading sometimes to http502.
As our customers still want this export feature even with thousands of rows, i have decided trying to go into a direction.
In order to still enable EXPORting data (to CSV), i would like to use combination of Serenity internal services.

a/ Add a button on grid allowing to export (batch and delayed) data
Once clicked a job is scheduled (saving definition of ListRequest settings into a JobQueue table holding list of /Services to call with associated parameters.
the request retrieveing all the data matching parameters is not executed immediately for better UserExperience, but user is receiving a notification his request has been taken into consideration.

b/ Based on submitted jobs (saved into a table), i would like to execute /List + associated parameters and produce a CSV file

c/ Once created the link to the generated CSV is then send (using MailJobQueue Service) to the user.

>> anyone did always started such a scenario ? i think it's pretty common when manipulating thousands/millions of records (it cannot be produced online but delayed) ?

So far i am stuck in ListRequest parameters being deserialized from DB and trying to dynamically call the appropriate /List url.

Thanks for any idea or help achieving this.

**Comment by edwardch** [comment]
Hey @kilroyFR, I'd also thought of a similar idea and got a little stuck at the ListRequest too. I use Hangfire for background jobs, which says best practice is to [keep arguments small and simple](https://docs.hangfire.io/en/latest/best-practices.html#make-job-arguments-small-and-simple), so parameterising the request is probably not the best idea.

If you can share what you have tried so far I wouldn't mind taking a look to see if I could contribute to some solution. My guess is we'd have to save the list request in a table...

**Reply by edwardch** [reply]
What I'll try is to first save the request in a table when the user requests a download, like so:

```cs
// Serialize request object as string
var requestAsString = Serenity.JSON.Stringify(request);
// Save request to requeet table
var yourQueuedRequest = InsertRequestToTable(requestAsString);
// Add your job to the queue
EnqueueDownloadAndEmailJob(yourQueuedRequest.ID);
```

Then to retrieve it in the background job, you could use something like the following:

```cs
// Get the request string
var requestAsString = YourMethodToGetRequestFromDatabase(backgroundJobId);
// Set list request
var listRequestObject = Serenity.JSON.Parse<ListRequest>(requestAsString);
// Perform Excel list request, save and email to user
```
